flume-ng-configuration/src/main/java/org/apache/flume/conf/FlumeConfiguration.java

```diff
@@ -348,75 +381,135 @@ private boolean isValid() {
+      if (LOGGER.isDebugEnabled() && LogPrivacyUtil.allowLogPrintConfig()) {
+        LOGGER.debug("Post validation configuration for {}", agentName);
+        LOGGER.debug(getPostvalidationConfig());
       }
 
       return true;
     }
 
-    private ChannelType getKnownChannel(String type) {
-      ChannelType[] values = ChannelType.values();
-      for (ChannelType value : values) {
-        if (value.toString().equalsIgnoreCase(type)) return value;
+    private void runFiltersThroughConfigs() {
+      runFiltersOnContextMaps(
+          sourceContextMap,
+          channelContextMap,
+          sinkContextMap,
+          sinkGroupContextMap
+      );
+    }
 
-        String channel = value.getChannelClassName();
+    private void runFiltersOnContextMaps(Map<String, Context>... maps) {
+      for (Map<String, Context> map: maps) {
+        for (Context c : map.values()) {
+          for (Entry<String, String> k : c.getParameters().entrySet()) {
+            c.put(k.getKey(), filterValue(k.getValue()));
+          }
+        }
+      }
+    }
 
-        if (channel != null && channel.equalsIgnoreCase(type)) return value;
+    private void createConfigFilters() {
+      for (String name: configFilterSet) {
+        Context context = configFilterContextMap.get(name);
+        ComponentConfiguration componentConfiguration = configFilterConfigMap.get(name);
+        try {
+
+          if (context != null) {
+            ConfigFilter configFilter = ConfigFilterFactory.create(
+                name, context.getString(BasicConfigurationConstants.CONFIG_TYPE)
+            );
+            configFilter.initializeWithConfiguration(context.getParameters());
+            configFiltersInstances.add(configFilter);
+          } else if (componentConfiguration != null) {
+            ConfigFilter configFilter = ConfigFilterFactory.create(
+                componentConfiguration.getComponentName(), componentConfiguration.getType()
+            );
+            configFiltersInstances.add(configFilter);
+          }
+        } catch (Exception e) {
+          LOGGER.error("Error while creating config filter {}", name, e);
+        }
+      }
+    }
 
+    private String filterValue(String value) {
+      for (ConfigFilter configFilter : configFiltersInstances) {
+        //JAVA EL expression style ${myFilterName['my_key']} or
+        //JAVA EL expression style ${myFilterName["my_key"]} or
+        //JAVA EL expression style ${myFilterName[my_key]}
+        List<String> delimiters = Arrays.asList("'", "\"", "");
+        for (String delimiter: delimiters) {
+          String prefix = "${" + configFilter.getName() + "[" + delimiter;
+          String suffix = delimiter + "]}";
+          boolean matchesWithDelimiter = value.startsWith(prefix) && value.endsWith(suffix);
+
+          if (matchesWithDelimiter) {
+            String filteredValue = configFilter.filter(value.substring(
+                configFilter.getName().length() + delimiter.length() + 3,
```

Comment:
This can be simplified to `prefix.length()`

Reply:
BTW, why not regexp with capture group and back reference for the delimiter?

Reply:
will change to regex and make it more flexible. I chose not to use regex, because it is always slower than simple string operations. Fortunately, it has to run only when config changes and only as many times as many lines the config has. A user has to have a large configuration to feel the difference, so It is ok to change it to regex.